gramps_webapi/api/resources/filters.py

```diff
@@ -0,0 +1,298 @@
+            if filter_rule[key["name"]] == rule_class.__name__:
+                rule_instance = rule_class
+                break
+        if rule_instance is None:
+            abort(400)
+        filter_args = []
+        if key["values"] in filter_rule:
+            filter_args = filter_rule[key["values"]]
+        filter_regex = False
+        for rkey in ["r", "regex"]:
+            if rkey in filter_rule:
+                filter_regex = filter_rule[rkey]
+        filter_object.add_rule(rule_instance(filter_args, use_regex=filter_regex))
+    return filter_object
+
+
+def apply_filter(db_handle: DbReadBase, args: Dict, namespace: str) -> List[Handle]:
+    """Apply an existing or dynamically defined filter."""
+    filters.reload_custom_filters()
+    if args.get("filter"):
+        for filter_class in filters.CustomFilters.get_filters(namespace):
+            if args["filter"] == filter_class.get_name():
+                return filter_class.apply(db_handle)
+        abort(400)
+
+    try:
+        filter_parms = QueryFilterSchema().load(json.loads(args["rules"]))
+    except json.JSONDecodeError:
+        abort(400)
+    except ValidationError:
+        abort(400)
+
+    filter_object = build_filter(filter_parms, namespace, keys="short")
+    return filter_object.apply(db_handle)
+
+
+class RuleSchema(Schema):
+    """Structure for a rule."""
+
+    name = fields.Str(required=True, validate=validate.Length(min=1))
+    values = fields.List(fields.Raw, required=False)
+    regex = fields.Boolean(required=False, missing=False)
+
+
+class FilterSchema(Schema):
+    """Structure for a filter."""
+
+    name = fields.Str(required=True, validate=validate.Length(min=1))
+    comment = fields.Str(required=False)
+    function = fields.Str(
+        required=False,
+        missing="and",
+        validate=validate.OneOf(["and", "or", "xor", "one"]),
+    )
+    invert = fields.Boolean(required=False, missing=False)
+    rules = fields.List(fields.Nested(RuleSchema))
+
+
+class QueryRuleSchema(Schema):
+    """Structure for validating a rule embedded in query parms."""
+
+    n = fields.Str(required=True, validate=validate.Length(min=1))
+    v = fields.List(fields.Raw, required=False)
+    r = fields.Boolean(required=False, missing=False)
+
+
+class QueryFilterSchema(Schema):
+    """Structure for validating a filter embedded in query parms."""
+
+    f = fields.Str(
+        required=False,
+        missing="and",
+        validate=validate.OneOf(["and", "or", "xor", "one"]),
+    )
+    i = fields.Boolean(required=False, missing=False)
+    r = fields.List(fields.Nested(QueryRuleSchema))
+
+
+class FilterResource(ProtectedResource, GrampsJSONEncoder):
+    """Filter resource."""
+
+    @use_args(
+        {"filter": fields.Str(), "rule": fields.Str()},
+        location="query",
+    )
+    def get(self, args: Dict, namespace: str) -> Response:
+        """Get available custom filters and rules."""
```

Comment:
This is a loooong method. Possible to refactor? We can also do it later, but also for unit tests it's difficult having so long methods where we cannot test individual pieces.

Reply:
I'm not sure criteria for what is considered long. I guess the two functions it performs could be broken out and called separately.